File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        solve(res, nums, {}, 0);
        return res;

    }
    void solve(vector<vector<int>>& res, vector<int> nums, vector<int> v, int index){
        if(nums.size() == index){ 
            res.push_back(v);
            return;
        }
        v.push_back(nums[index]);
        solve(res, nums, v, index+1);
        v.pop_back();
        solve(res, nums, v, index+1);
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        backTrack(res, nums, 0, new ArrayList<>());
        return res;  
    }

    private void backTrack(List<List<Integer>> res, int[] nums, int index, List<Integer> list){
        if(index == nums.length){
            res.add(new ArrayList<>(list));
            return;
        }
        else if(index > nums.length) return;
        list.add(nums[index]);
        backTrack(res, nums, index+1, list);
        list.remove(list.size()-1);
        backTrack(res, nums, index+1, list);
    }
}